### PR TITLE
Changing trace/open gadget to send absolute file path with symlink resolution in the events

### PR DESCRIPTION
**Pull request by slashben**
# Trace/open gadget to resolve full file paths and symlinks

Open trace gadget has been using the file name from the user-space side in the system call. 

The open file name argument can be:
1. Relative path: in this case, the observability tool must be able to point to the full path that was open
2. Symlink: the observability tool doesn't need to resolve symlinks (since it can be done outside of its scope). However, this creates a burden on the event receiver side. Therefore it is better systemwide to implement it on the event source side. 

In this PR we are implementing a real file path resolution within the trace/open gadget which uses the file description and kernel side file structure to resolve the full file path including symlinks and relative paths.

## How to use

This implementation simply adds the resolution to the existing gadget, therefore everyone who uses this gadget will automatically receive this improvement.

## Testing done

Added unit tests to cover both cases.

### CODE REVIEW

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c
Comment (alban) [inline pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c:129]: I agree with @eiffel-fl, we should use a different field for the given path and the resolved path.

pkg/gadgets/trace/open/types/types.go could have both `Path` and `FullPath`. One of them could have the column attribute "hide" if it is too long to print on one line.

**Comment by slashben** [comment]
> I am nonetheless wondering if we should not keep both information, _i.e._ the path which was opened, _e.g._ a relative one, and the underlying corresponding path, _e.g._ the absolute path corresponding.

@eiffel-fl great comment and I have to tell you that I have been thinking about it. These were the options I saw:
1. The gadget adds `cwd` to the event. This enables the event user to calculate the absolute path using the `filename` and `cwd`.
2. Resolve the full path (as in the current version) on the Kernel side and use it to the event as `filename`
3. Send both the `filename` from the system call and the full resolved path (as in option 2)

These were the factors for my decision:
* Sending two path-type strings in a single event inflates the event record. In general, paths are limited by PATH_MAX (4k) and in our case 256 bytes. AFAIK we have a hard time creating events larger than 512, no? This means that options 1 and 3 are creating a load on the ring buffer and the receiver side.
* `cwd` is not a very useful part of the event record, the only reason to have it is to reconstruct the absolute path. If I want to track the `cwd` I will follow the corresponding system call since the applications are not changing it so many times. Sending it in every event is suboptimal from a memory footprint perspective.
* Option number 1 does not resolve symbolic links. It is a big minus.
* Sending the original system call `filename` value helps to give a full picture of application behavior, and option 2 does not cover it.

Maybe it is due to my specific application needs, but I think that keeping the event pipeline memory friendly and the need for symbolic link resolution is more important than the other factors.

WDYT?

File: pkg/gadgets/common/filesystem.h
Comment (slashben) [reply pkg/gadgets/common/filesystem.h:182]: Agreed, changing it.

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c
Comment (slashben) [reply pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c:129]: @alban , see my comments below. Don't you think it will create pressure on the ring buffer and event structures? This would effectively double the size of an event record. Besides I am not sure about how useful the `Path` would be in this case.

**Comment by alban** [comment]
@slashben Thanks for explaining your considerations. Some additional points:

- I think it is better to resolve the paths in eBPF rather than userspace because:
  - By the time the event reaches userspace, the filesystem might have changed and the path resolution might be different. Even though it is not race-free in eBPF, it is still more likely to be correct.
  - We are trying to add support for containerized gadgets (#1722). If we reimplement trace-open as a containerized gadget, it's easier if most of the features are implemented in eBPF
- The limit to 512 bytes can be solved by allocating the event struct in a hash map indexed by the pid instead of the stack. There is an example of code pattern in [trace_exec.bpf.c#L90-L95](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/gadgets/trace_exec.bpf.c#L90-L95).
- If size of the event struct is a concern, we could have one field with a size limit such as 256 or more and add the two strings one after another. There is an example of this with args of execve in [execsnoop.h#L28](https://github.com/inspektor-gadget/inspektor-gadget/blob/3992c6dc120aaa0969c7f4e8bb35e1c095bee9f2/pkg/gadgets/trace/exec/tracer/bpf/execsnoop.h#L28) and [execsnoop.bpf.c#L110-L111](https://github.com/inspektor-gadget/inspektor-gadget/blob/3992c6dc120aaa0969c7f4e8bb35e1c095bee9f2/pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c#L110-L111)
- If memory or cpu performance becomes a problem, it is possible to enable/disable features in eBPF with parameters and `spec.RewriteConstants`. Example in [tracer.go#L210-L222](https://github.com/inspektor-gadget/inspektor-gadget/blob/3992c6dc120aaa0969c7f4e8bb35e1c095bee9f2/pkg/gadgets/profile/tcprtt/tracer/tracer.go#L210-L222) and [tcprtt.bpf.c#L18-L26](https://github.com/inspektor-gadget/inspektor-gadget/blob/3992c6dc120aaa0969c7f4e8bb35e1c095bee9f2/pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c#L18-L26)
  - We've got a user request for cwd for trace-exec (#1700) but not for trace-open yet. I wonder if we could get some numbers showing how much the performance is impacted or not by those additional fields.

File: pkg/gadgets/common/filesystem.h
Comment (slashben) [reply pkg/gadgets/common/filesystem.h:106]: Same here, I don't claim that I understand the code completely, but the structure of the `for` loop is looking for path components, not character level parsing.

**Comment by slashben** [comment]
@alban @eiffel-fl , I have changed the gadget to send the original path in the `Path` field (as before) and use `FullPath` for the resolved path. 
This is what you had in mind?

File: pkg/gadgets/trace/open/tracer/tracer_test.go
Comment (alban) [inline pkg/gadgets/trace/open/tracer/tracer_test.go:84]: Comment to be updated

File: pkg/gadgets/trace/open/tracer/tracer_test.go
Comment (alban) [inline pkg/gadgets/trace/open/tracer/tracer_test.go:257]: Tests are executed in parallel (`t.Parallel()` line 364) and the path `/tmp/test` is reused by several tests to create a file, create a symlink, or delete it. We need to use different paths in different tests to avoid race conditions.

This problem was not introduced in this PR, but I just noticed it.

**Review by alban — COMMENTED** [review]
This looks good! Just a couple of comments in the tests.

I didn't review pkg/gadgets/common/filesystem.h. Waiting from @mauriciovasquezbernal about it.

File: pkg/gadgets/trace/open/tracer/tracer_test.go
Comment (slashben) [reply pkg/gadgets/trace/open/tracer/tracer_test.go]: done

File: pkg/gadgets/trace/open/tracer/tracer_test.go
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/open/tracer/tracer_test.go:284]: Nice!

**Review by mauriciovasquezbernal — COMMENTED** [review]
Some initial comments. Will do a deeper pass soon.

File: pkg/gadgets/common/filesystem.h
Comment (mauriciovasquezbernal) [reply pkg/gadgets/common/filesystem.h:182]: This applies to few other places as well.

Comment (slashben) [reply pkg/gadgets/common/filesystem.h:182]: @mauriciovasquezbernal do you have a [built-in Clang formatting style](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#basedonstyle) you are using?

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.h
Comment (slashben) [reply pkg/gadgets/trace/open/tracer/bpf/opensnoop.h:27]: I think I understand what you mean in general, but I am not sure about the practice. Do you mean that if the last member of the C struct is the `full_name`, since it is a null-terminated string it can be truncated in the ring-buffer? Is this working with giving in the last argument `bpf_perf_event_output` less than `sizeof(struct event)`?

File: pkg/gadgets/common/filesystem.h
Comment (mauriciovasquezbernal) [reply pkg/gadgets/common/filesystem.h]: Unfortunately we haven't defined it yet -> https://github.com/inspektor-gadget/inspektor-gadget/issues/718. Now that I think better, you can just ignore this comment given that we don't have a style defined and requiring it is a non-sense.

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.h
Comment (mauriciovasquezbernal) [reply pkg/gadgets/trace/open/tracer/bpf/opensnoop.h:27]: The last argument to `bpf_perf_event_output` should be a variable that takes the return value of `bpf_probe_read_user_str` into consideration. The [trace exec](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c) gadget has a more complex example of this as in that case multiple strings are saved at the end.

File: pkg/gadgets/common/filesystem.h
Comment (mauriciovasquezbernal) [inline pkg/gadgets/common/filesystem.h:180]: nit: please remove trailing white spaces.

File: pkg/gadgets/common/filesystem.h
Comment (mauriciovasquezbernal) [inline pkg/gadgets/common/filesystem.h:201]: We use vmlinux.h that marks all structs with [preserve_access_index](https://github.com/alban/inspektor-gadget/blob/d135fa553e5a6957b9aee2274762db55b36db517/pkg/amd64/vmlinux/vmlinux.h#L5), so we don't need to make this manually. We can also use `get_path_str(&file->f_path)` directly to avoid creating the temp variable.

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c:26]: Remove `struct args_t` from opensnoop.h since it's not needed anymore after this change.

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c:132]: I know some people prefer to have the constant first in a comparison, but all our code uses the other way around. Please swap them.

**Review by mauriciovasquezbernal — COMMENTED** [review]
It's looking really good. Some minor comments around.

I'm a bit concerned about the performance impact of this, I think we should disable it by default. (@alban do you have comments on this?) You can get some inspiration about how to make it from https://github.com/inspektor-gadget/inspektor-gadget/pull/1700. But basically you have to:
- Introduce a new parameter (gadget.go). I'd suggest `full-path`
- Introduce a new global constant on the eBPF program and only get that field when it's true.
- Set the constant to true when the above parameter is true
- Enable the feature in integration tests.

Please also put a sentence in docs/gadgets/trace/open.md describing this new field (and how to enable it with the parameter introduced above).

Once that's complete, I'll approve. Thanks!

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.h
Comment (slashben) [reply pkg/gadgets/trace/open/tracer/bpf/opensnoop.h]: @mauriciovasquezbernal check out the code I have pushed, I have changed `full_fname` to be at PATH_MAX size and also putting variable size (depending on the string size) on the ring buffer.

File: pkg/gadgets/common/filesystem.h
Comment (slashben) [reply pkg/gadgets/common/filesystem.h:180]: done

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c
Comment (slashben) [reply pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c]: OK

File: pkg/gadgets/common/filesystem.h
Comment (slashben) [reply pkg/gadgets/common/filesystem.h]: gotcha

**Comment by mauriciovasquezbernal** [comment]
@slashben it looks good! Please organize the commits then I'll make a final pass.

**Comment by slashben** [comment]
@mauriciovasquezbernal I have reorganized the commits

**Comment by slashben** [comment]
```       cols := types.GetColumns()
          col, ok := cols.GetColumn("fullPath")
          if ok {
               col.GetAttributes().Visible = true
          }
```
@eiffel-fl  
Shouldn't this be conditional based on the `t.config.FullPath` flag?

File: pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c
Comment (slashben) [reply pkg/gadgets/trace/open/tracer/bpf/opensnoop.bpf.c:152]: Credits to @mauriciovasquezbernal

File: pkg/gadgets/common/filesystem.h
Comment (slashben) [reply pkg/gadgets/common/filesystem.h:53]: See the global variable declaration in line 49, it is in use. Since this is not my code, I'd prefer not to touch it :(

Comment (mauriciovasquezbernal) [reply pkg/gadgets/common/filesystem.h:53]: Yes, it can't be removed.

**Comment by mauriciovasquezbernal** [comment]
> I just would like to have the `fullPath` colum being printed when using the corresponding flag:

This can be done by adding `columnTags:"param:full-path` in pkg/gadgets/trace/open/types/types.go. 

```
FullPath string      `json:"fullPath,omitempty" column:"fullPath,minWidth:24,width:32" columnTags:"param:full-path"`
```

```bash
$ go run -exec "sudo -E" ./cmd/ig/. trace open --full-path
INFO[0000] Experimental features enabled                
CONTAINER                               PID        COMM             FD   ERR PATH                                     FULLPATH                                
minikube-containerd                     12139      kubelet          25   0   /proc/290884/net/dev                     /proc/290884/net/dev                    
minikube-containerd                     12139      kubelet          25   0   /sys/fs/cgroup/cpu,cpuacct/system.slice… /sys/fs/cgroup/cpu,cpuacct/system.slice…
minikube-containerd                     12139      kubelet          25   0   /proc/290884/fd                          /proc/290884/fd                         
minikube-containerd                     12139      kubelet          25   0   /proc/290884/limits                      /proc/290884/limits                     
minikube-containerd                     12139      kubelet          25   0   /sys/fs/cgroup/cpu,cpuacct/kubepods/bes… /sys/fs/cgroup/cpu,cpuacct/kubepods/bes…
minikube-containerd                     12139      kubelet          32   0   /sys/fs/cgroup/cpu,cpuacct/kubepods/bes… /sys/fs/cgroup/cpu,cpuacct/kubepods/bes…
minikube-containerd                     12139      kubelet          25   0   /sys/fs/cgroup/cpu,cpuacct/kubepods/bes… /sys/fs/cgroup/cpu,cpuacct/kubepods/bes…
minikube-containerd                     12139      kubelet          25   0   /proc/1310/net/dev                       /proc/1310/net/dev                      
minikube-containerd                     12139      kubelet          25   0   /sys/fs/cgroup/cpu,cpuacct/kubepods/bes… /sys/fs/cgroup/cpu,cpuacct/kubepods/bes…
minikube-containerd                     12139      kubelet          25   0   /proc/1310/fd                            /proc/1310/fd                           
minikube-containerd                     12139      kubelet          25   0   /proc/1310/limits                        /proc/1310/limits                       
minikube-containerd                     12139      kubelet          25   0   /sys/fs/cgroup/cpu,cpuacct/kubepods/bes… /sys/fs/cgroup/cpu,cpuacct/kubepods/bes…
^Cmvb@mvb-desktop:~/kinvolk/ebpf/inspektor-gadget$ 
```

**Review by mauriciovasquezbernal — APPROVED** [review]
LGTM. Thanks @slashben for your contribution. 

I added `columnTags:"param:full-path"` to types.go, fixed a commit message and pushed to your branch. I'll merge once CI is green.

**Comment by alban** [comment]
FYI, this PR made the `trace open` gadget 5x slower to start, even when `--full-path` is not given: about 1 second instead of 176ms. See comment generated by the bot at the bottom of the commit:
https://github.com/inspektor-gadget/inspektor-gadget/commit/204c97f7a72661c5fa08f07a4e4730244b6615d8#commitcomment-121440900

image

[source of the graph](https://inspektor-gadget.github.io/ig-benchmarks/dev/bench/index.html)

If someone is interested in profiling it, there is documentation how to do that step by step:
https://github.com/inspektor-gadget/inspektor-gadget/blob/main/docs/devel/CONTRIBUTING.md#benchmarks

My guess is it's the ebpf verifier (in kernel) and the cilium/ebpf BTF relocations (in userspace) and there is not much we can do about it.

**Pull request merged**
